modules/tensor_mechanics/src/materials/ComputeEigenstrainFromInitialStress.C

```diff
@@ -0,0 +1,72 @@
+/****************************************************************/
+/* MOOSE - Multiphysics Object Oriented Simulation Environment  */
+/*                                                              */
+/*          All contents are licensed under LGPL V2.1           */
+/*             See LICENSE for full restrictions                */
+/****************************************************************/
+
+#include "ComputeEigenstrainFromInitialStress.h"
+#include "RankTwoTensor.h"
+#include "Function.h"
+
+template <>
+InputParameters
+validParams<ComputeEigenstrainFromInitialStress>()
+{
+  InputParameters params = validParams<ComputeEigenstrainBase>();
+  params.addClassDescription("Computes an eigenstrain from an initial stress");
+  params.addParam<std::vector<FunctionName>>(
+      "initial_stress",
+      "A list of functions describing the initial stress.  If provided, there "
+      "must be 9 of these, corresponding to the xx, yx, zx, xy, yy, zy, xz, yz, "
+      "zz components respectively.  If not provided, all components of the "
+      "initial stress will be zero");
+  params.addParam<std::string>("base_name",
+                               "The base_name for the elasticity tensor that will be "
+                               "used to compute strain from stress.  Do not provide "
+                               "any base_name if your elasticity tensor does not use "
+                               "one.");
+  params.set<bool>("incremental_form") = true;
+  return params;
+}
+
+ComputeEigenstrainFromInitialStress::ComputeEigenstrainFromInitialStress(
+    const InputParameters & parameters)
+  : ComputeEigenstrainBase(parameters),
+    _base_name(isParamValid("base_name") ? getParam<std::string>("base_name") + "_" : ""),
+    _elasticity_tensor(getMaterialPropertyByName<RankFourTensor>(_base_name + "elasticity_tensor"))
+{
+  if (_incremental_form == false)
+    mooseError("ComputeEigenstrainFromInitialStress: incremental_form must be set to true");
+
+  const std::vector<FunctionName> & fcn_names(
+      getParam<std::vector<FunctionName>>("initial_stress"));
+  const unsigned num = fcn_names.size();
+
+  if (num != LIBMESH_DIM * LIBMESH_DIM)
+    mooseError("ComputeEigenstrainFromInitialStress: ",
+               LIBMESH_DIM * LIBMESH_DIM,
+               " initial stress functions must be provided.  You supplied ",
+               num,
+               "\n");
+
+  _initial_stress_fcn.resize(num);
+  for (unsigned i = 0; i < num; ++i)
+    _initial_stress_fcn[i] = &getFunctionByName(fcn_names[i]);
+}
+
+void
+ComputeEigenstrainFromInitialStress::computeQpEigenstrain()
+{
+  if (_t_step == 1)
```

Comment:
I'm mildly concerned about what happens if the elasticity tensor changes a great deal in the first timestep, say with a large temperature spike or lots of damage, but I'm not sure how relevant of a corner case this is.  Perhaps it might be worth running the calculation in initialQpProperties but then saving the eigenstrain value calculated in `_t_step --1`?  Or maybe a message in the class description to use care if the elasticity tensor changes drastically at the simulation start?

Reply:
I toyed with the idea of running the eigenstrain calculation in initQp (for other reasons actually).  It won't work without a lot of bother, since the elasticity tensor is not defined then (it's not stateful).  I don't think it's worth going to a great deal of effort at the moment before we really have a use-case where this matters.  I'll certainly include a message in the doco string - thanks.

If you can see a better way of doing this, please tell me!

Reply:
I added a big note regarding this to the .md file

Reply:
I guess the concern here is if the tensor changes between time 0 and time 1.  Let's go ahead and merge this code as it is, but I'd like to see if there's some way we can figure out how to compute the eigenstrain using the elasticity tensor as it is at time 0, so that it would always do the right thing.